### PR TITLE
CosmosDB Mongo vCore - Adding HNSW vector index support 

### PR DESCRIPTION
**Pull request by aayush3011**
Read more about [Vector Search in CosmosDB Mongo vCore](https://learn.microsoft.com/en-us/azure/cosmos-db/mongodb/vcore/vector-search).

Adding HNSW vector index type support for CosmosDB Mongo vCore vector search. I have also added more properties needed for vector search in the cosmos db configs.

Updated the test cases as well.

### CODE REVIEW

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchConfigOptions.cs [hunk 1]
```diff
@@ -13,4 +13,42 @@ public class CosmosDBSearchConfigOptions
...
     // This integer is the number of clusters that the inverted file (IVF) index uses to group the vector data
     public int NumLists { get; set; } = 1;
+
+    /// <summary>
+    /// Type of vector index to create.
+    ///     Possible options are:
+    ///         - vector-ivf
+    ///         - vector-hnsw
+    /// </summary>
+    public AzureCosmosDBVectorSearchType Kind { get; set; } =
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchConfigOptions.cs:23]: Can we update the samples for the usage of this?

Also, README file for the rag-cosmos samples - https://github.com/Azure/azure-functions-openai-extension/blob/main/samples/rag-cosmosdb/README.md

Let me know if you face any issues with running the samples, bit tricky to run dotnet-isolated samples while consuming recent changes when they have not been published to nuget yet.

Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchConfigOptions.cs]: Updated in the host files, and the readme file as well.

File: samples/rag-cosmosdb/python/local.settings.json [hunk 2]
```diff
@@ -5,9 +5,10 @@
...
     "CosmosDBMongoVCoreConnectionString": "",
-    "AZURE_OPENAI_ENDPOINT": "https://<resource-name>.openai.azure.com/",
-    "CHAT_MODEL_DEPLOYMENT_NAME": "gpt-3.5-turbo",
-    "EMBEDDING_MODEL_DEPLOYMENT_NAME": "text-embedding-ada-002",
-    "PYTHON_ISOLATE_WORKER_DEPENDENCIES": "1"
+    "AZURE_OPENAI_ENDPOINT": "",
+    "CHAT_MODEL_DEPLOYMENT_NAME": "gpt-35-turbo",
+    "EMBEDDING_MODEL_DEPLOYMENT_NAME": "vector_search_task",
+    "PYTHON_ISOLATE_WORKER_DEPENDENCIES": "1",
+    "AZURE_OPENAI_KEY": ""
```
Comment (manvkaur) [inline samples/rag-cosmosdb/python/local.settings.json:12]: don't modify any default values in the local.settings.json of samples unless you are introducing any new setting

File: samples/rag-cosmosdb/python/host.json [hunk 3]
```diff
@@ -9,8 +9,19 @@
...
-        "numLists": 1
+        "numLists": 1,
+        "kind": "VectorIVF",
+        "similarity": "Cosine",
+        "numberOfConnections": 16,
+        "efConstruction": 64,
+        "efSearch": 40,
+        "databaseName": "functionsDatabase",
+        "collectionName": "functionsCollection",
+        "indexName": "functionsIndex",
```
Comment (manvkaur) [inline samples/rag-cosmosdb/python/host.json:22]: comment in discussion for collection name, database name, index name and connection string at other place.

File: samples/rag-cosmosdb/README.md [hunk 4]
```diff
@@ -22,39 +22,53 @@ Once you have an Cosmos DB resource, you can run the sample by following these s
...
+               "kind": "vector-hnsw",
+               "similarity": "COS",
+               "numberOfConnections": 16,
+               "efConstruction": 64,
+               "efSearch": 40
+     }
+   }
+           }
+       }
+   }
```
Comment (manvkaur) [inline samples/rag-cosmosdb/README.md:47]: problem with indentation and number of braces for this json

File: samples/rag-cosmosdb/README.md [hunk 5]
```diff
@@ -22,39 +22,53 @@ Once you have an Cosmos DB resource, you can run the sample by following these s
...
+   ```
+
+   `VectorSearchDimensions` is length of the embedding vector. [The dimensions attribute has a minimum of 2 and a maximum of 2000 floating point values each](https://learn.microsoft.com/azure/cosmos-db/mongodb/vcore/vector-search#create-an-vector-index-using-ivf). By default, the length of the embedding vector will be 1536 for text-embedding-ada-002.
+
+   `NumLists` is the number of clusters that the inverted file (IVF) uses to group the vector data, as mentioned [here](https://learn.microsoft.com/azure/cosmos-db/mongodb/vcore/vector-search#create-an-vector-index-using-ivf). By default, the number of clusters will be 1.
```
Comment (manvkaur) [inline samples/rag-cosmosdb/README.md:53]: new settings should also be explained here and default values for them if unspecified

**Review by manvkaur — CHANGES_REQUESTED** [review]

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 6]
```diff
@@ -16,10 +16,11 @@ sealed class CosmosDBSearchProvider : ISearchProvider
...
     readonly ILogger logger;
-    readonly int vectorSearchDimensions = 1536;
-    readonly int numLists = 1;
-    const string defaultSearchIndexName = "openai-index";
-    const string databaseName = "openai-database";
+    readonly IOptions<CosmosDBSearchConfigOptions> cosmosDBSearchConfigOptions;
+    MongoClient mongoClient;
+    String databaseName;
+    String collectionName;
+    String indexName;
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: Updated the code.

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 7]
```diff
@@ -45,26 +46,49 @@ public OutputDocument(string title, string text)
...
         {
             throw new ArgumentNullException(nameof(loggerFactory));
         }
 
+        if (string.IsNullOrEmpty(cosmosDBSearchConfigOptions.Value.ConnectionString))
+        {
+            throw new ArgumentNullException(
+                "Connection String property in the cosmosDBSearchConfigOptions can not be empty or null."
+            );
+        }
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: Updated the code, as per discussion.

File: samples/rag-cosmosdb/demo.http [hunk 8]
```diff
@@ -3,25 +3,25 @@
 POST http://localhost:7071/api/IngestFile
 Content-Type: application/json
```
Comment (manvkaur) [inline samples/rag-cosmosdb/demo.http:5]: what do you suggest?

File: src/Functions.Worker.Extensions.OpenAI/Search/ConnectionInfo.cs [hunk 9]
```diff
@@ -14,15 +14,24 @@ public class ConnectionInfo
...
+    /// The name of the database which has the collection.</param>
+    /// </param>
     /// <param name="collectionName">
     /// The name of the collection or table to search or store.</param>
     /// </param>
     /// <param name="credentials">
     /// Credentials for authenticating with the search provider.</param>
     /// </param>
-    public ConnectionInfo(string connectionName, string collectionName, string? credentials)
+    public ConnectionInfo(
```
Comment (manvkaur) [inline src/Functions.Worker.Extensions.OpenAI/Search/ConnectionInfo.cs:26]: As discussed, please take any additional properties from host file for now, modifying this file will impact all other search providers too, like Kusto and AI Search

Also, this is causing the failures in build, please check it for errors

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 10]
```diff
@@ -16,10 +17,11 @@ sealed class CosmosDBSearchProvider : ISearchProvider
...
     readonly ILogger logger;
-    readonly int vectorSearchDimensions = 1536;
-    readonly int numLists = 1;
-    const string defaultSearchIndexName = "openai-index";
-    const string databaseName = "openai-database";
+    readonly IOptions<CosmosDBSearchConfigOptions> cosmosDBSearchConfigOptions;
+    readonly ConcurrentDictionary<string, MongoClient> cosmosDBClients = new();
+    string databaseName = "openai-functions-database";
+    string collectionName = "openai-functions-collection";
+    string indexName = "openai-functions-index";
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:24]: why do we need default values when this is mandatory for users to supply?

Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: Removed them.

File: src/Functions.Worker.Extensions.OpenAI/Search/ConnectionInfo.cs (hunk 9 above)
Comment (aayush3011) [reply src/Functions.Worker.Extensions.OpenAI/Search/ConnectionInfo.cs]: I have moved it to the config file for now, but we should add it to the ConnectionInfo class.

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 11]
```diff
@@ -102,56 +120,38 @@ public async Task<SearchResponse> SearchAsync(SearchRequest request)
...
-        }
-
-        string connectionString = this.configuration.GetValue<string>(request.ConnectionInfo.ConnectionName);
-        MongoClient mongoClient = this.GetMongoClient(connectionString, request.ConnectionInfo.ConnectionName);
-
         try
         {
-            IMongoCollection<BsonDocument> collection = mongoClient.GetDatabase(databaseName).GetCollection<BsonDocument>(request.ConnectionInfo.CollectionName ?? defaultSearchIndexName);
-
+            MongoClient cosmosClient = cosmosDBClients.GetOrAdd(
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:125]: us this qualification for cosmosDBClients

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 12]
```diff
@@ -102,56 +120,38 @@ public async Task<SearchResponse> SearchAsync(SearchRequest request)
...
+                request.ConnectionInfo.ConnectionName,
+                _ => CreateMongoClient(request.ConnectionInfo.ConnectionName)
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:127]: use this qualification for CreateMongoClient

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 13]
```diff
@@ -73,19 +81,29 @@ public CosmosDBSearchProvider(IConfiguration configuration, ILoggerFactory logge
...
+    )
     {
-        if (document.ConnectionInfo is null)
-        {
-            throw new ArgumentNullException(nameof(document.ConnectionInfo));
-        }
-        string connectionString = this.configuration.GetValue<string>(document.ConnectionInfo.ConnectionName);
+        MongoClient cosmosClient = cosmosDBClients.GetOrAdd(
+            document.ConnectionInfo!.ConnectionName,
+            _ => CreateMongoClient(document.ConnectionInfo.ConnectionName)
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:91]: use this qualification for cosmosDBClients and CreateMongoClient

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 14]
```diff
@@ -73,19 +81,29 @@ public CosmosDBSearchProvider(IConfiguration configuration, ILoggerFactory logge
...
+        );
 
-        MongoClient mongoClient = this.GetMongoClient(connectionString, document.ConnectionInfo.ConnectionName);
+        this.databaseName = cosmosDBSearchConfigOptions.Value.DatabaseName;
+        this.collectionName = document.ConnectionInfo.CollectionName;
+        this.indexName = cosmosDBSearchConfigOptions.Value.IndexName;
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:96]: use this qualification for cosmosDBSearchConfigOptions

[the remark this answers is not in the document]
Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:96]: For Mongo vCore Search the collection name and index name will be different, you can have multiple vector indexes in the same collection.

[the remark this answers is not in the document]
Comment (manvkaur) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: 1. this.indexName is unused and looks like cosmosDBSearchConfigOptions.Value.IndexName is directly being used to pass the indexName. There should be consistency with the other properties assignment and usage
2. Also the member configuration is unused

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs [hunk 15]
```diff
@@ -73,19 +81,29 @@ public CosmosDBSearchProvider(IConfiguration configuration, ILoggerFactory logge
...
+        this.collectionName = document.ConnectionInfo.CollectionName;
+        this.indexName = this.cosmosDBSearchConfigOptions.Value.IndexName;
+        this.CreateVectorIndexIfNotExists(cosmosClient);
 
-        this.CreateVectorIndexIfNotExists(mongoClient, document.ConnectionInfo.CollectionName ?? defaultSearchIndexName);
+        await this.UpsertVectorAsync(cosmosClient, document);
+    }
 
-        await this.UpsertVectorAsync(mongoClient, document);
+    MongoClient CreateMongoClient(String connectionName)
```
Comment (manvkaur) [inline src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs:102]: should be `string`

File: src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: Updated the code to use this.indexName.

[the remark this answers is not in the document]
Comment (manvkaur) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: what about configuration property? it is still unused?

[the remark this answers is not in the document]
Comment (aayush3011) [reply src/WebJobs.Extensions.OpenAI.CosmosDBSearch/CosmosDBSearchProvider.cs]: Configuration property is not unused. I have this code.

`this.indexName = this.cosmosDBSearchConfigOptions.Value.IndexName;`

**Comment by manvkaur** [comment]
Please resolve the conflicts and also update the host file for `samples/rag-cosmosdb/javascript`

nodejs folder has been split into typescript and javascript

File: eng/ci/public-build-java.yml [hunk 16]
```diff
@@ -60,7 +60,7 @@ extends:
 
     stages:
       - stage: BuildJava
-        displayName: 'Build Java'
+        displayName: "Build Java"
         jobs:
```
Comment (manvkaur) [inline eng/ci/public-build-java.yml:64]: changes in this file are not required

**Comment by manvkaur** [comment]
it seems there are lot of unintended changes in this PR, merge conflicts pushed unresolved.

**Pull request closed without merging**